### PR TITLE
Ft social login 159054011

### PR DESCRIPTION
**Pull request by Tittoh**
**What does this PR do?**
Implement social login via Facebook and Google

**Description of Task to be completed?**
- Authenticate user
- Add user to database
- Give the user a JWT token

**How should this be manually tested?**
1. Clone this branch
2. 
``` 
cd ah-jarvis

create and activate a virtual env 

pip install -r requirements.txt

coverage run manage.py test

python manage.py run server

```
3. Register a new application at [Facebook App Creation](http://developers.facebook.com/setup/)
4. Add the App ID and SECRET from [here](https://developers.facebook.com/apps/284874268731044/settings/basic/) to your `.env` file as shown in `env.example`
5. Generate a test user [here](https://developers.facebook.com/apps/284874268731044/roles/test-users/)
6. Edit the test user: `change permissions this test user granted to app` and add `email` `public_profile` as the parameters
screen shot 2018-08-03 at 06 30 15

7. From the edit button select `Get an access token for this test user` to get an `access_token` to be used in Postman

8. Go to [Google console](https://console.developers.google.com/apis/dashboard) to create a new app
9. Add the App ID and SECRET for the app to your `.env` file
10. Get a token from [google OAuth playground](https://developers.google.com/oauthplayground/) and follow the steps. In step one, expand `Apps Script API v1` and select `https://www.googleapis.com/auth/userinfo.email` then authorise API. 
11. Select the account to authenticate and exchange authorisation code for tokens in step 2 and an `access_token` is given to be used in Postman.

12. On Postman, go to `http://127.0.0.1:8000/api/convert_token/ ` to convert the access_token to an authentication token. Add this to the body

```
{
	"access_token":"<enter access_token from facebook here>",
	"provider":"facebook"
}
```
or
```
{
	"access_token":"<enter access_token from google here>",
	"provider":"google-oaut2"
}
```

- the response should look like this 
```
{
    "user": {
        "email": "jake@jake.jake",
        "username": "Jacob",
        "token": "xxxxxxxxxxxxxxxx"
    }
}
```

**Any background context you want to provide?**
None

**What are the relevant pivotal tracker stories?**
[#159054011](https://www.pivotaltracker.com/story/show/159054011)

### CODE REVIEW

File: authors/apps/authentication/views.py [hunk 1]
```diff
@@ -73,3 +75,48 @@ def update(self, request, *args, **kwargs):
...
 
+class SocialSignUp(CreateAPIView):
+    permission_classes = (AllowAny,)
+    renderer_classes = (UserJSONRenderer,)
+    serializer_class = UserSerializer
+
+    def create(self, request,  *args, **kwargs):
+        # Get access token and create user or authenticate
+
+        provider = request.data['provider']
```
Comment (dwarukira) [inline authors/apps/authentication/views.py]: Handle error if this is not provided. You can use a serializer.

File: authors/apps/authentication/views.py [hunk 2]
```diff
@@ -73,3 +75,48 @@ def update(self, request, *args, **kwargs):
...
+
+        authed_user = request.user if not request.user.is_anonymous else None
+
+        strategy = load_strategy(request)
+
+        backend = load_backend(strategy=strategy, name=provider, redirect_uri=None)
+
+        if isinstance(backend, BaseOAuth2):
+            # Grab the access_token
+            token = request.data['access_token']
```
Comment (dwarukira) [inline authors/apps/authentication/views.py]: Handle error if this is not provided to.

**Review by dwarukira — CHANGES_REQUESTED** [review]
I like the work so far. A few changes will improve error handling.

File: authors/apps/authentication/views.py (hunk 1 above)
Comment (Tittoh) [reply authors/apps/authentication/views.py]: Thanks for noticing that. I'll work on it.

File: authors/apps/authentication/views.py (hunk 2 above)
Comment (Tittoh) [reply authors/apps/authentication/views.py]: noted.

**Comment by jacksono** [comment]
@Tittoh fix the failing build in Travis. You can check the Job log on Travis to see where its failing

**Comment by jacksono** [comment]
@Tittoh there are merge conflicts you need to resolve

**Review by jacksono — APPROVED** [review]

**Comment by jacksono** [comment]
Merging this in pending twitter approval

**Pull request merged**